### DISCUSSION TITLE
Add _after_ready() to ensure that logic runs when all nodes in scenetree are ready

### DISCUSSION THREAD

**Discussion by azur-wolve** [godotengine/godot-proposals#10889] [Engine Core]
Godot aims for intuitiveness and usability.

One of the problems that newcomers always face is the initialization order of nodes regarding READY.
It has been said to be found counter-intuitive multiple times.
_(as for the rest of the tree creation, the steps go from up hierarchy to down. Is the ready step that only goes from down to up)_

The children come first as its expected that nodes higher in the tree hierarchy use nodes that are lower and not vice-versa, which is a good practice.

However there are cases where is needed to check something on an ancestor from a descendant.

Currently `_ready()` is the last "step" before the first game loop iteration

if there were an `_after_ready()` could solve some of that, so its ensured that all nodes have been initialized and are ready with `_init()` and `_ready()` respectively.

Also its handy as we move nodes around the hierarchy when trying new things on scene prototyping stages, so we don't have to worry where they are moved to, and change the code every time, or being limited to avoid changing it.
All that regarding to the initialization step when launching a scene.

Right now to have such behavior would need something like:

```gdscript

func _ready():
    # some ready stuff
    _after_ready.call_deferred()

func _after_ready():
    # whatever
```

Maybe that could throw some problem even?

**Comment by sockeye-d** [comment]
If you need to wait for a parent to be ready, you can do
```gdscript
await get_parent().ready
```
Or if you need to wait for the whole tree to be ready:
```gdscript
await get_tree().root.ready
```
I think this is a nice idea but it's really easy to work around it

**Reply by azur-wolve** [reply]
what idk is if its recommendable to populate the stack with multiple awaits
maybe someone can clarify

**Reply by OffsetMOSFET** [reply]
It's an await that happens once during setup. It should be fine.

Also, to add to sockeye-d's example, if you can't guarantee a node's load order (sibling/cousin node), I recommend just checking is_node_ready() first, otherwise you might be awaiting after the ready signal has fired.

**Reply by keyzeuh** [reply]
I made a pull request about this: https://github.com/godotengine/godot/pull/108413

**Comment by dalexeev** [comment]
Next time someone asks for `_after_after_ready()`? I think it's redundant. We already have `_enter_tree()` which is called from parent to child and `_ready()` which is called from child to parent. The fact that you need an additional callback probably indicates a problem with the architecture of your project and a violation of the intended relationship between parent and child nodes ("signal up, method down").

**Reply by tomsterBG** [reply]
Factorio has three of them haha!
https://lua-api.factorio.com/latest/auxiliary/data-lifecycle.html
<img width="2274" height="330" alt="image" src="https://github.com/user-attachments/assets/6c00d2b1-4f5d-4efe-8bd4-50f5933324b9" />
Again:
<img width="2274" height="330" alt="image" src="https://github.com/user-attachments/assets/f69560fa-7afd-4d08-9912-5573abe0d48e" />
And again, but this is different, kinda:
<img width="3171" height="774" alt="image" src="https://github.com/user-attachments/assets/8656333e-e06b-45ba-95de-0d9243b7096a" />

I can see how this is very useful for a modding platform where people just can't agree on which mod to load when. Mod dependencies help as you can depend on a mod so it loads first and your script in the lifecycle will always run after theirs, but sometimes you want to change all mods, even ones you don't know or have never seen. There's no way around this, but the closest thing is using `xxxxx-final-fixes.lua` which works 99% of the time, unless someone else has a mod that uses it incorrectly and causes your mod to modify everything and then theirs adds something that your mod misses. However i still see no reason for the middle `xxxxx-updates.lua` because you would have already made your stuff in `xxxxxx.lua` and other mods can depend on you to change your code. Perhaps its for the rare dual mod case where two mods (pipe visualizer and belt visualizer for example) make compatibilities together and one has to depend on the other, but both can't as it would cause a circle that is ambiguous. So modB would depend on modA and update it in `xxxxx.lua`, but if modA wants to update modB it has to do it in `xxxxx-updates.lua`.

Not to mention the stages are divided and you can't access or do everything (as opposed in Godot) which means Factorio kinda needs this a lot more than Godot, but they still only use three levels.

So if Godot needs anything after `_ready()`, it probably can be worked around or you can add a reverse `_after_ready()` that goes from parent to children like how `_enter_tree()` works. Otherwise it's pretty useless for us other than having it as a plugin. Or maybe a different implementation, a dynamic `_after_ready()` which can be told how many passes it will do and in what direction (up/down) and it will give an argument telling you which pass it's on. That would be cool, but also overkill so i'm waiting for a real reason to justify it.

In Roblox i had many instantiation problems, but those are usually worked around with wait loops because it may take a while until the action is done. Or when possible i used events instead to make it more efficient. Still it wasn't some determinism-requiring order (Factorio is deterministic so yet another reason for their solution), it was just me waiting for the server to do its job which can't be ensured to complete during the loading stage, and using something like `_ready()` assumes you think it's fast enough to complete now and wouldn't be some network async call, that's for the game loop.
